### ISSUE TITLE
Should We Rewrite MyBatis Generator?

### ISSUE THREAD

**Issue by topsoft-support** [mybatis/generator#481]
I looked at the source code, if you want to support the generation of kotlin code, you need to change a lot of places, so let's invite more capable people to add it.

**Comment by jeffgbutler** [comment]
tl;dr Requires a rewrite, might not be worth the effort

I've thought about this a lot.

As you may have noticed, there is not good separation between the "model" that is generated and the rendering of Java code. This means that it is not really possible to just plug in a different renderer. It will take a rewrite to accomplish the goal of different language support.

Some years ago, I started working on a rewrite here: https://github.com/jeffgbutler/mybatis-generator2. You can see some of my goals for a new generator there. As I progressed in the rewrite, I realized there were actually two major issues: the generator code needs to be improved for sure, but the **generated** code was also not very good.

I decided to work on the second problem first because it is far more impactful. The result of that effort is https://github.com/mybatis/mybatis-dynamic-sql. Using this library dramatically reduces the amount of generated code - which is a good thing. But more importantly, it allows us to generate code that is far more capable than the original code created by the generator. In a perfect world, I would remove the code generators for all the old style Java and XML code that was produced from the original generator and just generate code for MyBatis Dynamic SQL. But I know that's a drastic step. Over time I hope that all users will adopt this new paradigm and we can eventually remove the old code generators. That would also make a rewrite easier.

You can see an example of what my idea of generated Kotlin code might look like here: https://github.com/jeffgbutler/mybatis-kotlin-examples. In that repo look at example05 - it's my idea of what generated Kotlin code would look like.

With MyBatis Dynamic SQL, the generated Java code is much smaller and there is no generated XML. The generated Java code is small enough that it might be possible to run it through a Java to Kotlin converter. This is where I start to wonder if the rewrite is worth the effort.

So, bottom line, I'm not sure what the next steps should be and I would like to solicit some feedback from the community. If we were to rewrite the generator I think it would have these impacts:

1. It would only generate code for MyBatis Dynamic SQL
1. It would break the existing plugin mechanism - because plugins are tied to Java code now, and `IntrospectedTable` is a giant object with too many responsibilities - it needs to go away!
1. The new generator configuration files would not be the same as the current generator - maybe not even XML
1. Migration would be required both on the generator configuration, and for code that uses the generated code

What do you think - is it worth it?

**Comment by topsoft-support** [comment]
I think you're right. Today I tried to add the code that generates kotlin, add  DefaultKotlinFormatter.java, GeneratedKotlinFile.java, modify MyBatisGenerator.java and Plugin, and so on. Later I found that I couldn't go on. From the beginning, the design was based on Java, although plug-ins could be used , but always Java-oriented. It will take a lot of effort to refactor the code to support kotlin , which seems not very cost-effective. At present, in my project, I use Java and kotlin mixed code, the Example generated by generator uses Java, and the other uses kotlin. 
Thank you for your contribution to generator

**Comment by jeffgbutler** [comment]
Have you tried the runtime for MyBatis Dynamic SQL? Less ugly Java code in your beautiful Kotlin project 😃

**Comment by topsoft-support** [comment]
The MyBatis Dynamic SQL looks good. It's more powerful than generator's Example. The java code in kolin project really doesn't look very good. But this is the safest solution at the moment.Now I want to run my project in a production environment with code I'm more familiar with and more controllable. I will focus on MyBatis Dynamic SQL in the future, if it really works well and replace them at last.

**Comment by jeffgbutler** [comment]
@topsoft-support - see #515. I've made the changes necessary to support Kotlin code generation in the basic library. It turned out not to be that difficult. I did not try to make an abstract DOM that was shared between Java and Kotlin. Instead I added a new simple Kotlin DOM. I think this is better anyway.

So I will close this issue as we now have a path for Kotlin in the existing generator.
